Update Node Binary Installation Guide (#874)
maticnetwork/matic-docs: docs/maintain/validate/run-validator.md

---
id: run-validator
title: Run a Validator Node
description: Use package to set up your validator node on the Polygon Network.
keywords:
  - docs
  - matic
  - polygon
  - binary
  - node
  - validator
  - sentry
image: https://matic.network/banners/matic-network-16x9.png
---
import useBaseUrl from '@docusaurus/useBaseUrl';

> 

Reviewer: ```suggestion
```

:::caution

> This guide only applies to Bor and Heimdall version >= 0.3.0. For older versions, please follow the guides in 

Reviewer: ```suggestion
This guide only applies to Bor and Heimdall version >= 0.3.0. For older versions, please follow
```

> [Full Node Binaries (Legacy)](run-validator-binaries.md).

Reviewer: ```suggestion
the legacy [Validator guide](run-validator-binaries.md).
```

:::

This guide will walk you through running a Polygon validator node from packages.

For system requirements,
follow the [Validator Node System Requirements](validator-node-system-requirements.md) guide.

## Overview

To get to a running validator node, conduct the following in this **exact sequence of steps**:

> You will run into configuration issues if these steps are performed out of sequence.
> It is important to keep in mind that a sentry node must always be set up before the validator node.

1. Prepare two machines, one for the sentry node and one for the validator node.
2. Install the Heimdall and Bor binaries on the sentry and validator machines.
3. Set up the Heimdall and Bor service files on the sentry and validator machines.
4. Set up the Heimdall and Bor services on the sentry and validator machines.
5. Configure the sentry node.
6. Start the sentry node.
7. Configure the validator node.
8. Set the owner and signer keys.
9. Start the validator node.
10. Check node health with the community.

## Installing packages

> There are three different ways to install heimdall and bor binaries: [installing packages](#install-with-packages-recommended), [installing using ansible](#install-with-ansible), and [building from source code](#install-from-source-code).

Reviewer: ```suggestion
There are three different ways to install heimdall and bor binaries: [installing packages via utility scripts](#install-with-packages-recommended), [installing via Ansible](#install-with-ansible), and [building from source code](#install-from-source-code).
```

  children:
    sentry:
      hosts:
        xxx.xxx.xx.xx: # <----- Add IP for sentry node
        xxx.xxx.xx.xx: # <----- Add IP for second sentry node (optional)
    validator:
      hosts:
        xxx.xxx.xx.xx: # <----- Add IP for validator node
```

Example:

```yml
all:
  hosts:
  children:
    sentry:
      hosts:
        188.166.216.25:
    validator:
      hosts:
        134.209.100.175:
```

Check that the remote sentry machine is reachable. On the local machine, run:

```sh
$ ansible sentry -m ping
```

You should get this as output:

```sh
xxx.xxx.xx.xx | SUCCESS => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python3"
    },
    "changed": false,
    "ping": "pong"
}
```

Do a test run of the node setup:

```sh
ansible-playbook playbooks/network.yml --extra-var="bor_branch=v0.3.0 heimdall_branch=v0.3.0" --list-hosts
```

This will be the output:

```sh
playbook: playbooks/network.yml
  pattern: ['all']
  host (1):
    xx.xxx.x.xxx
```

Once the setup is complete, you will see a message of completion on the terminal.

:::note

If you run into an issue and would like to start over, run:

```sh
ansible-playbook playbooks/clean.yml
```

:::

> 

Reviewer: ```suggestion
```

### Install from source code

#### Prerequisites

* Two machines — one [sentry](../glossary#sentry) and one [validator](../glossary#validator).
* `build-essential` installed on both the sentry and the validator machines.

  To install:

  ```sh
  sudo apt-get install build-essential
  ```

* Go 1.17 installed on both the sentry and the validator machines.

  To install:

  ```sh
  wget https://gist.githubusercontent.com/ssandeep/a6c7197811c83c71e5fead841bab396c/raw/go-install.sh
  bash go-install.sh
  sudo ln -nfs ~/.go/bin/go /usr/bin/go
  ```

* RabbitMQ installed on both the sentry and the validator machines.
  See [Downloading and Installing RabbitMQ](https://www.rabbitmq.com/download.html).

> 

Reviewer: ```suggestion
```

  Example: `private_peer_ids = "0ee1de0515f577700a6a4b6ad882eff1eb15f066"`.

* `prometheus` — set the value to `true` to enable the Prometheus metrics. Example: `prometheus = true`.
* `max_open_connections` — set the value to `100`. Example: `max_open_connections = 100`.

Save the changes in `config.toml`.

### Configure Bor

Create a new file `static-nodes.json` with command `mkdir -p /lib/var/bor/data/bor/ && vi /lib/var/bor/data/bor/static-nodes.json`.

In `static-nodes.json`, add the following:

```
[
  "<replace with enode://validator_machine_enodeID@validator_machine_ip:30303>"
]
``` 

To get the node ID of Bor on the validator machine:

1. Log into the validator machine.
2. Run `bor bootnode -node-key /var/lib/bor/data/bor/nodekey -dry-run`.

Example output: `"enode://410e359736bcd3a58181cf55d54d4e0bbd6db2939c5f548426be7d18b8fd755a0ceb730fe5cf7510c6fa6f0870e388277c5f4c717af66d53c440feedffb29b4b@134.209.100.175:30303"`.

Example content of `static-nodes.json`:
```
[
  "enode://410e359736bcd3a58181cf55d54d4e0bbd6db2939c5f548426be7d18b8fd755a0ceb730fe5cf7510c6fa6f0870e388277c5f4c717af66d53c440feedffb29b4b@134.209.100.175:30303"
]
```

Save the changes in `static-nodes.json`.

> 

Reviewer: ```suggestion
```

### Configuring a firewall

The sentry machine must have the following ports open to the world `0.0.0.0/0`:

* `26656`- Your Heimdall service will connect your node to other nodes Heimdall service.

* `30303`- Your Bor service will connect your node to other nodes Bor service.

* `22`- For the validator to be able to ssh from wherever they are.

> 

Reviewer: ```suggestion
```

## Migration

Bor 0.3.0 and Heimdall 0.3.0 uses new CLIs and path standards. It is recommended to set up everything from a new machine.

> However, if you still want to perform upgrade on existing node, you need to follow one-time migration steps 

Reviewer: ```suggestion
However, if you still want to perform an upgrade on existing node, you need to follow the one-time migration steps 
```

    sudo mv $(go env GOPATH)/bin/bor ./backup/go/bin
    sudo mv $(go env GOPATH)/bin/heimdalld ./backup/go/bin
    sudo mv $(go env GOPATH)/bin/heimdallcli ./backup/go/bin
    sudo mv $(go env GOPATH)/bin/bridge ./backup/go/bin
    ```

- Move old service files
    ```shell
    sudo mv /etc/systemd/system/bor.service ./backup
    sudo mv /etc/systemd/system/heimdalld.service ./backup
    sudo mv /etc/systemd/system/heimdalld-rest-server.service ./backup
    sudo mv /etc/systemd/system/heimdalld-bridge.service ./backup
    ```

- Migrate heimdall and bor directory to `/var/lib`
    ```shell
    sudo mv ~/.heimdalld /var/lib/heimdall
    sudo mv ~/.bor /var/lib/bor
    ```
    In case data copying is too slow or original data folder is mounted on a different device, you can create symlinks
    ```shell
    sudo ln -nfs ~/.heimdalld /var/lib/heimdall
    sudo ln -nfs ~/.bor /var/lib/bor
    ```

- Copy configurations in `node/bor/start.sh` to bor service file `/lib/systemd/system/bor.service`. Note that some 
  flags are renamed in the new CLI, you can find the documentation for new CLI [here](https://github.com/maticnetwork/bor/tree/master/docs/cli).

> 

Reviewer: ```suggestion
```

> ## Configure service files for bor and heimdall

Reviewer: ```suggestion
## Configure service files for Bor and Heimdall
```

    
    - In the service file, set `--chain` to `mainnet` or `mumbai` accordingly, example:

      ```
          --chain=mumbai \
      ```

    - Set service user to a user with restricted permission, example:

      ```
      User=ubuntu
      ```

  Save the changes in `/lib/systemd/system/heimdalld.service`.

- Make sure the chain is set correctly in `/lib/systemd/system/bor.service` file. Open the file with following command `sudo vi /lib/systemd/system/bor.service`

    - In the service file, set `--chain` to `mainnet` or `mumbai` accordingly, example:

      ```
          -chain=mumbai \
      ```

    - Set service user to a user with restricted permission, example:

      ```
      User=ubuntu
      ```

  Save the changes in `/lib/systemd/system/bor.service`.

## Starting the Sentry Node

You will first start the Heimdall service. Once the Heimdall service syncs, you will start the Bor service.

### Reload service files

Reloading service files to make sure all changes to service files are loaded correctly.

> ```

Reviewer: ```suggestion
```sh
```

In `heimdall-config.toml`, change the following:

* `eth_rpc_url` — an RPC endpoint for a fully synced Ethereum mainnet node or testnet node,
  i.e Infura. `eth_rpc_url =<insert Infura or any full node RPC URL to Ethereum>`

Example: `eth_rpc_url = "https://nd-123-456-789.p2pify.com/60f2a23810ba11c827d3da642802412a"`

Save the changes in `heimdall-config.toml`.

### Configuring Bor

Create a new file `static-node.json` with command `mkdir -p /var/lib/bor/data/bor && vi /var/lib/bor/data/bor/static-nodes.json`.

In `static-nodes.json`, change the following:

To get the node ID of Bor on the sentry machine:

1. Log in to the sentry machine.
2. Run `bor bootnode -node-key /var/lib/bor/data/bor/nodekey -dry-run`.

Example output: `"enode://a8024075291c0dd3467f5af51a05d531f9e518d6cd229336156eb6545581859e8997a80bc679fdb7a3bd7473744c57eeb3411719b973b2d6c69eff9056c0578f@188.166.216.25:30303"`.

Example content of `static-nodes.json`:
```
[
  "enode://a8024075291c0dd3467f5af51a05d531f9e518d6cd229336156eb6545581859e8997a80bc679fdb7a3bd7473744c57eeb3411719b973b2d6c69eff9056c0578f@188.166.216.25:30303"
]
```

Save the changes in `static-nodes.json`.

> 

Reviewer: ```suggestion
```

---
id: run-validator
title: Run a Validator Node
description: Use package to set up your validator node on the Polygon Network.
keywords:
  - docs

>   - matic

Reviewer 2: ```suggestion
  - polygon wiki
```

Reviewer: we've been using "matic" as a generic tag

## Overview

To get to a running validator node, conduct the following in this **exact sequence of steps**:

> You will run into configuration issues if these steps are performed out of sequence.
> It is important to keep in mind that a sentry node must always be set up before the validator node.

1. Prepare two machines, one for the sentry node and one for the validator node.
2. Install the Heimdall and Bor binaries on the sentry and validator machines.
3. Set up the Heimdall and Bor service files on the sentry and validator machines.
4. Set up the Heimdall and Bor services on the sentry and validator machines.
5. Configure the sentry node.
6. Start the sentry node.
7. Configure the validator node.
8. Set the owner and signer keys.
9. Start the validator node.
10. Check node health with the community.

## Installing packages

There are three different ways to install heimdall and bor binaries: [installing packages via utility scripts](#install-with-packages-recommended), [installing via Ansible](#install-with-ansible), and [building from source code](#install-from-source-code).

### Install with packages (recommended)
#### Prerequisites

> * Two machines — one [sentry](/docs/validate/glossary#sentry) and one [validator](/docs/validate/glossary#validator).

Reviewer 2: ```suggestion
* Two machines — one [sentry](/maintain/glossary.md#sentry) and one [validator](/maintain/glossary.md#validator).
```

* Bash is installed on both the sentry and the validator machines.

* RabbitMQ installed on both the sentry and the validator machines.
  See [Downloading and Installing RabbitMQ](https://www.rabbitmq.com/download.html).

#### Heimdall

- Install the default latest version of sentry for mainnet:

    ```shell
    curl -L https://raw.githubusercontent.com/maticnetwork/install/main/heimdall.sh | bash
    ```

    or install a specific version, node type (`sentry` or `validator`), and network (`mainnet` or `testnet`). All release versions can be found on 
    [Heimdall GitHub repository](https://github.com/maticnetwork/heimdall/releases).

    ```shell
    curl -L https://raw.githubusercontent.com/maticnetwork/install/main/heimdall.sh | bash -s -- <version> <network> <node_type>
    # Example:
    # curl -L https://raw.githubusercontent.com/maticnetwork/install/main/heimdall.sh | bash -s -- v0.3.0 mainnet sentry
    ```

#### Bor

> - Install the default latest version of sentry for mainnet:

Reviewer 2: ```suggestion
- Install the default latest version of sentry for Mainnet:
```

    ```shell
    curl -L https://raw.githubusercontent.com/maticnetwork/install/main/bor.sh | bash
    ```

    or install a specific version,  node type (`sentry` or `validator`), and network (`mainnet` or `testnet`). All release versions could be found on 
    [Bor Github repository](https://github.com/maticnetwork/bor/releases).

    ```shell
    curl -L https://raw.githubusercontent.com/maticnetwork/install/main/bor.sh | bash -s -- <version> <network> <node_type>

    # Example:
    # curl -L https://raw.githubusercontent.com/maticnetwork/install/main/bor.sh | bash -s -- v0.3.0 mainnet sentry
    ```

### Install with ansible

#### Prerequisites

> * Three machines — one local machine on which you will run the Ansible playbook; two remote machines — one [sentry](../glossary#sentry) and one [validator](../glossary#validator).

Reviewer 2: ```suggestion
* Three machines — one local machine on which you will run the Ansible playbook; two remote machines — one [sentry](/maintain/glossary.md#sentry) and one [validator](/maintain/glossary.md#validator).
```

* On the local machine, [Ansible](https://www.ansible.com/) installed.
* On the local machine, [Python 3.x](https://www.python.org/downloads/) installed.
* On the remote machines, make sure Go is *not* installed.
* On the remote machines, your local machine's SSH public key is on the remote machines to let Ansible connect to them.

> * We have Bloxroute available as a relay network. If you need a gateway to be added as your Trusted Peer please contact [Delroy on Discord](http://delroy/#0056).

Reviewer 2: ```suggestion
* We have **Bloxroute** available as a relay network. If you need a gateway to be added as your **Trusted Peer**, please contact [Delroy on Discord](http://delroy/#0056).
```

        xxx.xxx.xx.xx: # <----- Add IP for second sentry node (optional)
    validator:
      hosts:
        xxx.xxx.xx.xx: # <----- Add IP for validator node
```

Example:

```yml
all:
  hosts:
  children:
    sentry:
      hosts:
        188.166.216.25:
    validator:
      hosts:
        134.209.100.175:
```

Check that the remote sentry machine is reachable. On the local machine, run:

```sh
$ ansible sentry -m ping
```

You should get this as output:

```sh
xxx.xxx.xx.xx | SUCCESS => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python3"
    },
    "changed": false,
    "ping": "pong"
}
```

Do a test run of the node setup:

```sh
ansible-playbook playbooks/network.yml --extra-var="bor_branch=v0.3.0 heimdall_branch=v0.3.0" --list-hosts
```

This will be the output:

```sh
playbook: playbooks/network.yml
  pattern: ['all']
  host (1):
    xx.xxx.x.xxx
```

Once the setup is complete, you will see a message of completion on the terminal.

:::note

If you run into an issue and would like to start over, run:

```sh
ansible-playbook playbooks/clean.yml
```

:::

### Install from source code

#### Prerequisites

> * Two machines — one [sentry](../glossary#sentry) and one [validator](../glossary#validator).

Reviewer 2: ```suggestion
* Two machines — one [sentry](/maintain/glossary.md#sentry) and one [validator](/maintain/glossary.md#validator).
```

* `build-essential` installed on both the sentry and the validator machines.

  To install:

  ```sh
  sudo apt-get install build-essential
  ```

* Go 1.17 installed on both the sentry and the validator machines.

  To install:

  ```sh
  wget https://gist.githubusercontent.com/ssandeep/a6c7197811c83c71e5fead841bab396c/raw/go-install.sh
  bash go-install.sh
  sudo ln -nfs ~/.go/bin/go /usr/bin/go
  ```

* RabbitMQ installed on both the sentry and the validator machines.
  See [Downloading and Installing RabbitMQ](https://www.rabbitmq.com/download.html).

#### Installing Heimdall

> [Heimdall](../validator/core-components/heimdall-chain.md) is the proof-of-stake verifier layer

Reviewer 2: ```suggestion
[Heimdall](/maintain/validator/core-components/heimdall-chain.md) is the proof-of-stake verifier layer
```

> responsible for checkpointing the representation of the Plasma blocks to the Ethereum mainnet.

Reviewer 2: ```suggestion
responsible for checkpointing the representation of the Plasma blocks to the Ethereum Mainnet.
```

Clone the [Heimdall repository](https://github.com/maticnetwork/heimdall/):

```sh
git clone https://github.com/maticnetwork/heimdall
```

Navigate to the correct [release version](https://github.com/maticnetwork/heimdall/releases):

```sh
git checkout RELEASE_TAG
```

where `RELEASE_TAG` is the tag of the release version that you install.

For instance:

```sh
git checkout v0.3.0
```

Once you are on the correct release, install Heimdall:

```sh
make install
```

Create symlinks:

```sh
sudo ln -nfs $(which heimdalld) /usr/local/bin/heimdalld
sudo ln -nfs $(which heimdallcli) /usr/local/bin/heimdallcli
```

#### Installing Bor

> [Bor](../../contribute/bor/) is the sidechain operator that acts as the block production layer,

Reviewer 2: ```suggestion
[Bor](/pos/bor/overview.md) is the sidechain operator that acts as the block production layer,
```

> which syncs with Heimdall to select block producers and verifiers for each [span](../glossary#span)

Reviewer 2: ```suggestion
which syncs with Heimdall to select block producers and verifiers for each [span](/maintain/glossary.md#span)
```

> and [sprint]((../glossary#sprint)).

Reviewer 2: ```suggestion
and [sprint]((/maintain/glossary.md#sprint)).
```

```sh
git clone https://github.com/maticnetwork/bor
```

Navigate to the correct [release version](https://github.com/maticnetwork/bor/releases):

```sh
git checkout RELEASE_TAG
```

where `RELEASE_TAG` is the tag of the release version that you install.

For instance:

```sh
git checkout v0.3.0
```

Install Bor:

```sh
make bor
```

Create symlinks:

```sh
sudo ln -nfs ~/bor/build/bin/bor /usr/local/bin/bor
```

### Check installation

- Check Heimdall installation

    ```shell
    heimdalld version --long
    ```

- Check Bor installation

    ```shell
    bor version
    ```

:::note

Before proceeding, Bor should be installed on both the sentry and validator machines.

:::

## Configuration

In this section, we will go through steps to initialize and customize configurations nodes. 

:::caution

Bor and Heimdall 0.3.0 uses standardized paths for configuration files and chain data. If you have existing 
config files and chain data on your node, please skip this section and jump directly to **[Migration](#upgrade-from-02x-to-03x) section** to learn about migrating configs and data to standardized file locations.

:::

### Configure Heimdall

> - Initialize heimdall configs

Reviewer 2: ```suggestion
- Initialize Heimdall configs
```

Example content of static node field in `/var/lib/bor/config.toml`:
```
[p2p]
    [p2p.discovery]
        static-nodes = ["enode://410e359736bcd3a58181cf55d54d4e0bbd6db2939c5f548426be7d18b8fd755a0ceb730fe5cf7510c6fa6f0870e388277c5f4c717af66d53c440feedffb29b4b@134.209.100.175:30303"]
```

Save the changes in `/var/lib/bor/config.toml`.

### Configuring a firewall

The sentry machine must have the following ports open to the world `0.0.0.0/0`:

* `26656`- Your Heimdall service will connect your node to other nodes Heimdall service.

* `30303`- Your Bor service will connect your node to other nodes Bor service.

* `22`- For the validator to be able to ssh from wherever they are.

## Upgrade from 0.2.x to 0.3.x

Bor 0.3.0 and Heimdall 0.3.0 uses new CLIs and path standards. It is recommended to set up everything from a new machine.
However, if you still want to perform an upgrade on existing node, you need to follow the one-time migration steps 
outlined below. If you are installing everything from a new machine, you can skip this section and continue to [Configure service files](#configure-service-files-for-bor-and-heimdall).

> - Stop existing heimdall and bor services

Reviewer 2: ```suggestion
- Stop existing Heimdall and Bor services
```

    sudo mv $(go env GOPATH)/bin/bor ./backup/go/bin
    sudo mv $(go env GOPATH)/bin/heimdalld ./backup/go/bin
    sudo mv $(go env GOPATH)/bin/heimdallcli ./backup/go/bin
    sudo mv $(go env GOPATH)/bin/bridge ./backup/go/bin
    ```

- Move old service files
    ```shell
    sudo mv /etc/systemd/system/bor.service ./backup
    sudo mv /etc/systemd/system/heimdalld.service ./backup
    sudo mv /etc/systemd/system/heimdalld-rest-server.service ./backup
    sudo mv /etc/systemd/system/heimdalld-bridge.service ./backup
    ```

- Migrate heimdall and bor directory to `/var/lib` and change ownership:
    ```shell
    sudo mv ~/.heimdalld /var/lib/heimdall
    sudo mv ~/.bor /var/lib/bor
    sudo chown -R heimdall /var/lib/heimdall
    sudo chown -R bor /var/lib/bor
    ```

    In case data copying is too slow or original data folder is mounted on a different device, you can create symlinks

    ```shell
    sudo chown -R heimdall ~/.bor
    sudo chown -R bor ~/.heimdalld
    sudo rm -rf /var/lib/heimdall
    sudo ln -nfs ~/.heimdalld /var/lib/heimdall
    sudo ln -nfs ~/.bor /var/lib/bor
    sudo chown -R heimdall /var/lib/heimdall
    sudo chown -R bor /var/lib/bor
    ```

> - Copy configurations in `node/bor/start.sh` to bor configuration file `/var/lib/bor/config.toml`. Note that some 

Reviewer 2: ```suggestion
- Copy configurations in `node/bor/start.sh` to Bor configuration file `/var/lib/bor/config.toml`. Note that some 
```

  flags are renamed in the new CLI, you can find the documentation for new CLI [here](https://github.com/maticnetwork/bor/tree/master/docs/cli) and sample configuration file in [launch repository](https://github.com/maticnetwork/launch).

>   You can use [this util script](https://github.com/maticnetwork/bor/blob/develop/scripts/getconfig.sh) to convert start.sh to a config.toml file on your host. Example usage:

Reviewer 2: ```suggestion
  You can use [this util script](https://github.com/maticnetwork/bor/blob/develop/scripts/getconfig.sh) to convert `start.sh` to a `config.toml` file on your host. Example usage:
```

  ```shell
  $ git clone https://github.com/maticnetwork/bor.git
  $ cd bor/scripts
  $ BOR_DIR=/var/lib/bor ./getconfig.sh
  * Path to start.sh: /home/ubuntu/node/bor/start.sh
  * Your validator address (e.g. 0xca67a8D767e45056DC92384b488E9Af654d78DE2), or press Enter to skip if running a sentry node: 0xca67a8D767e45056DC92384b488E9Af654d78DE2
  Thank you, your inputs are:
  Path to start.sh: /home/ubuntu/node/bor/start.sh
  Address: 0xca67a8D767e45056DC92384b488E9Af654d78DE2
  Path to the config file: /home/ubuntu/node/bor/start-config.toml
  ...

  $ sudo cp /home/ubuntu/node/bor/start-config.toml /var/lib/bor/config.toml
  $ sudo chown bor /var/lib/bor/config.toml
  ```

## Configure service files for Bor and Heimdall

> After successfully installing Bor and Heimdall through [packages](#install-with-packages-recommended), their service file could be found under `/lib/systemd/system`, and bor's config 

Reviewer 2: ```suggestion
After successfully installing Bor and Heimdall through [packages](#install-with-packages-recommended), their service file could be found under `/lib/systemd/system`, and Bor's config 
```

file could be found under `/var/lib/bor/config.toml`. 
You will need to check and modify these files accordingly.

    - In the service file, set `--chain` to `mainnet` or `mumbai` accordingly

  Save the changes in `/lib/systemd/system/heimdalld.service`.

- Make sure the chain is set correctly in `/var/lib/bor/config.toml` file. Open the file with following command `sudo vi /var/lib/bor/config.toml`

    - In the config file, set `chain` to `mainnet` or `mumbai` accordingly.

>     - To enable Archive mode you can optionally enable the following flags:

Reviewer 2: ```suggestion
    - To enable Archive mode, you can optionally enable the following flags:
```

```sh
heimdallcli generate-keystore ETHEREUM_PRIVATE_KEY
```

where

* ETHEREUM_PRIVATE_KEY — your Ethereum private key.

When prompted, set up a password to the keystore file.

This will generate a `UTC-<time>-<address>` keystore file.

Move the generated keystore file to the Bor configuration directory:

```sh
mv ./UTC-<time>-<address> /var/lib/bor/keystore/
```

### Add password.txt

Make sure to create a `password.txt` file then add the Bor keystore file password right in the
`/var/lib/bor/password.txt` file.

### Add your Ethereum address

Open for editing `vi /etc/matic/metadata`.

In `metadata`, add your Ethereum address. Example: `VALIDATOR_ADDRESS=0xca67a8D767e45056DC92384b488E9Af654d78DE2`.

Save the changes in `metadata`.

## Upgrade from 0.2.x to 0.3.x

Bor 0.3.0 and Heimdall 0.3.0 uses new CLIs and path standards. It is recommended to set up everything from a new machine.
However, if you still want to perform upgrade on existing node, you need to follow one-time migration steps 
outlined below. If you are installing everything from a new machine, you can skip this section and continue to [Configure service files](#configure-service-files-for-bor-and-heimdall-1).

> - Stop existing heimdall and bor services

Reviewer 2: ```suggestion
- Stop existing Heimdall and Bor services
```

    ```shell
    sudo service bor stop
    sudo service heimdalld stop
    sudo service heimdalld-rest-server stop
    sudo service heimdalld-bridge stop
    ```

- Create a backup folder in case something went wrong
    ```shell
    mkdir ./backup
    mkdir ./backup/bin
    mkdir -p ./backup/go/bin
    ```

- Move old binaries
    ```shell
    sudo mv /usr/bin/bor ./backup/bin
    sudo mv /usr/bin/heimdalld ./backup/bin
    sudo mv /usr/bin/heimdallcli ./backup/bin
    sudo mv /usr/bin/bridge ./backup/bin
    sudo mv $(go env GOPATH)/bin/bor ./backup/go/bin
    sudo mv $(go env GOPATH)/bin/heimdalld ./backup/go/bin
    sudo mv $(go env GOPATH)/bin/heimdallcli ./backup/go/bin
    sudo mv $(go env GOPATH)/bin/bridge ./backup/go/bin
    ```

- Move old service files
    ```shell
    sudo mv /etc/systemd/system/bor.service ./backup
    sudo mv /etc/systemd/system/heimdalld.service ./backup
    sudo mv /etc/systemd/system/heimdalld-rest-server.service ./backup
    sudo mv /etc/systemd/system/heimdalld-bridge.service ./backup
    ```

> - Migrate heimdall and bor directory to `/var/lib` and change ownership:

Reviewer 2: ```suggestion
- Migrate the Heimdall and Bor directory to `/var/lib` and change ownership:
```

  flags are renamed in the new CLI, you can find the documentation for new CLI [here](https://github.com/maticnetwork/bor/tree/master/docs/cli) and sample configuration file in [launch repository](https://github.com/maticnetwork/launch).

  You can use [this util script](https://github.com/maticnetwork/bor/blob/develop/scripts/getconfig.sh) to convert start.sh to a config.toml file on your host. Example usage:

  ```shell
  $ git clone https://github.com/maticnetwork/bor.git
  $ cd bor/scripts
  $ BOR_DIR=/var/lib/bor ./getconfig.sh
  * Path to start.sh: /home/ubuntu/node/bor/start.sh
  * Your validator address (e.g. 0xca67a8D767e45056DC92384b488E9Af654d78DE2), or press Enter to skip if running a sentry node: 0xca67a8D767e45056DC92384b488E9Af654d78DE2
  Thank you, your inputs are:
  Path to start.sh: /home/ubuntu/node/bor/start.sh
  Address: 0xca67a8D767e45056DC92384b488E9Af654d78DE2
  Path to the config file: /home/ubuntu/node/bor/start-config.toml
  ...

  $ sudo cp /home/ubuntu/node/bor/start-config.toml /var/lib/bor/config.toml
  $ sudo chown bor /var/lib/bor/config.toml
  ```

## Configure service files for bor and heimdall

> After successfully installing Bor and Heimdall through [packages](#install-with-packages-recommended), their service file could be found under `/lib/systemd/system`, and bor's config 

Reviewer 2: ```suggestion
After successfully installing Bor and Heimdall through [packages](#install-with-packages-recommended), their service file could be found under `/lib/systemd/system`, and Bor's config 
```

For snapshot download links, see [Polygon Chains Snapshots](https://snapshots.matic.today/).

:::

Start the Heimdall services:

```sh
sudo service heimdalld start
```

Check the Heimdall service logs:

```sh
journalctl -u heimdalld.service -f
```

Check the sync status of Heimdall:

```sh
curl localhost:26657/status
```

In the output, the `catching_up` value is:

* `true` — the Heimdall service is syncing.
* `false` — the Heimdall service is synced.

Wait for the Heimdall service to fully sync.

### Starting the Bor service

Once the Heimdall service on the validator machine syncs, start the Bor service on
the validator machine.

Start the Bor service:

```sh
sudo service bor start
```

Check the Bor service logs:

```sh
journalctl -u bor.service -f
```

## Health Checks with the Community

Now that your sentry and validator nodes are in sync and running, head over to
[Discord](https://discord.gg/polygon) and ask the community to health-check your nodes.

## Next Steps: Staking

Now that you have your sentry and validator nodes are health-checked, proceed to

> the [Staking](../validator/core-components/staking.md) guide to start backing the network.

Reviewer 2: ```suggestion
the [Staking](/maintain/validator/core-components/staking.md) guide to start backing the network.
```
